presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicBucketNodeMap.java

```diff
@@ -58,7 +58,19 @@ public DynamicBucketNodeMap(ToIntFunction<Split> splitToBucket, int bucketCount,
     @Override
     public Optional<InternalNode> getAssignedNode(int bucketedId)
     {
-        return Optional.ofNullable(bucketToNode.get(bucketedId));
+        if (!bucketToNodeAndCache.containsKey(bucketedId)) {
+            return Optional.empty();
+        }
+        return Optional.of(bucketToNodeAndCache.get(bucketedId).getInternalNode());
+    }
+
+    @Override
+    public Optional<Boolean> getAssignedNodeCache(int bucketedId)
```

Comment:
Maybe name this `isBucketCacheable`? Also, if bucket does not exist, I'm guessing we should also return `false`? An empty result is a bit confusing.